### PR TITLE
Ci rework gcc builds

### PR DESCRIPTION
**Pull request by scottwittenburg**
This PR reworks how docker images are built for gcc, clang, and intel builds done in github actions.  To dramatically reduce build times for the gcc builds, the plan going forward is to use pre-built Docker images from the E4S project (see "Minimal Images", described [here](https://e4s-project.github.io/download.html)) as a base image.  These images come with some spack packages already installed, but more importantly, installing ADIOS2 dependencies using the spack shipped in those images results in most dependencies getting installed from a release-specific E4S buildcache.  This is far faster than building those dependencies from sources.

For intel builds run in el8, since no buildcache is available to install ADIOS2 dependencies for that OS, the dependencies that can be installed from the package manager are installed, and options for missing dependencies are just turned off.

After this PR is merged, there are only two shell scripts used to build, tag, and push all the images used in gcc and intel builds on github actions.  To build gcc images (currently including gcc8, gcc9, gcc10, gcc11, and clang10), simply run `scripts/ci/images-v2/build-ubuntu.sh`.  On a fairly old linux machine with 32 hyperthreads, but with slow spinning disks, and with no caching, this now takes around 30 minutes.  To build, tag, and push images for icc and oneapi testing, run `scripts/ci/images-v2/build-el8.sh`.  This takes around 20 minutes, under the same conditions mentioned above.

This PR also renders some of the existing dockerfiles, scripts, cmake files, etc unused, and thus removes them from the repo.

fixes: #3611
fixes: #3612
fixes: #3613
fixes: #3614

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@vicentebolea The two failing tests on ubuntu/gcc10/mpich seem to have failed on master recently as well.  

There is also 1 configure warning on each of the four intel builds.  I think it's because those images are using the latest nightly version of cmake, which is warning about kwsys having too old of a minimum cmake version requirement.

Other than those issues, this is probably ready for you to review.

File: .github/workflows/everything.yml
Comment (vicentebolea) [inline .github/workflows/everything.yml:179]:
```suggestion
```

We probably do not need this anymore, it was for the GPU builds that could not run tests as we did not have machines with GPUs.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake:41]: We need to merge the files from cmake-v2 to the cmake folder giving priority to the files from cmake-v2.

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Review from the skies, I will review the rest tomorrow, running out of battery in the fly.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake
Comment (scottwittenburg) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake:41]: I think I'm missing something, do you mean we need that eventually?  Or as a part of this PR?  What is the goal there, just cleaning up old stuff we don't use anymore?

File: .github/workflows/everything.yml
Comment (scottwittenburg) [reply .github/workflows/everything.yml:179]: Ok, I can remove that.

File: .shellcheck_exclude_paths
Comment (scottwittenburg) [reply .shellcheck_exclude_paths:26]: I didn't like shellcheck suggestions, and I noticed there were a lot of exceptions, so I just created a few more.   I guess you're saying new scripts should be written in compliance with whatever shellcheck expects?

File: scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf
Comment (scottwittenburg) [reply scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf]: Ah, I misunderstood that, I'll make them all the same.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake
Comment (vicentebolea) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-clang10-mpi.cmake:41]: I was referring to doing it in this PR but yeah lets have this merge soon, lets do this in another PR.

File: .shellcheck_exclude_paths
Comment (vicentebolea) [reply .shellcheck_exclude_paths:26]: >  I guess you're saying new scripts should be written in compliance with whatever shellcheck expects?

Yes

File: scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf
Comment (scottwittenburg) [reply scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf]: I'm not familiar with the `--slave` thing, but I just tried what you suggested and got an error:

```
#0 32.48 update-alternatives: error: alternative g++ can't be slave of gcc: it is a master alternative
```

If it's important to use the `--slave` arguments, I can dig into this to try to figure how to make it work.  Unless you've already encountered it and know how to fix it.

Comment (scottwittenburg) [reply scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf]: Neither `gcc` nor `g++` can be a slave apparently, so afaict those have to be done in separate lines.  So we might as well just leave this as it is, at least imo.

Comment (vicentebolea) [reply scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-leaf]: I Tried `update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 100 --slave /usr/bin/g++ g++ /usr/bin/g++-11` in the ubuntu image and it seems to work.

I can see that Ubuntu does uses update-alterantives for the symlinks cc and c++ which then points to another gcc|g++ symlink which then point to the correct version of gcc|g++. The latter symlink is set when installing a gcc|g++ version (or here when the `alternative` is installed). I do not see a problem slaving the alternatives for the GCC dist.

**Comment by scottwittenburg** [comment]
@vicentebolea Since a lot of the work that went into this PR was exploratory at first, the commit history is far from clean.  I'm not sure it's worth the effort to try to break it up into logical chunks now, but I could easily collapse it into a single commit if you want.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake:38]: Why using make

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake:39]: Same as above

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc8-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc8-mpi.cmake:39]: Same

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc9-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc9-mpi.cmake:39]: Same

File: scripts/ci/gh-actions/run.sh
Comment (vicentebolea) [inline scripts/ci/gh-actions/run.sh:142]: remove comment

File: scripts/ci/images-v2/Dockerfile.ci-el8-intel
Comment (vicentebolea) [inline scripts/ci/images-v2/Dockerfile.ci-el8-intel:54]: This seems to risky, how about its latest release?

File: scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-gcc
Comment (vicentebolea) [inline scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-gcc:9]: Use slave

File: scripts/ci/setup-run/ci-ubuntu20.04-clang10.sh
Comment (vicentebolea) [inline scripts/ci/setup-run/ci-ubuntu20.04-clang10.sh:3]: Nice :)

File: scripts/ci/setup-run/ci-ubuntu20.04-gcc10.sh
Comment (vicentebolea) [inline scripts/ci/setup-run/ci-ubuntu20.04-gcc10.sh:3]: A single --set would be needed when using slave

File: scripts/ci/setup-run/ci-ubuntu20.04-gcc11.sh
Comment (vicentebolea) [inline scripts/ci/setup-run/ci-ubuntu20.04-gcc11.sh:3]: A single --set would be needed when using slave

File: scripts/ci/setup-run/ci-ubuntu20.04-gcc8.sh
Comment (vicentebolea) [inline scripts/ci/setup-run/ci-ubuntu20.04-gcc8.sh:3]: A single --set would be needed when using slave

File: scripts/ci/setup-run/ci-ubuntu20.04-gcc9.sh
Comment (vicentebolea) [inline scripts/ci/setup-run/ci-ubuntu20.04-gcc9.sh:3]: A single --set would be needed when using slave

File: source/adios2/engine/bp5/BP5Reader.cpp
Comment (vicentebolea) [inline source/adios2/engine/bp5/BP5Reader.cpp:388]: IMHO go ahead and remove every comment here, just remain: 

```
        for (auto &f : futures)
        {
            f.get();
        }
```

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Great job Scott, only cosmetic changes requested here. Sorry for me taking this long to review this PR, the reason is due to the size of changes and me being distracted with the AHM this week. Let me know

**Comment by vicentebolea** [comment]
We also are missing the clang6 build but we can provide it in a later PR

**Comment by vicentebolea** [comment]
> I'm not sure it's worth the effort to try to break it up into logical chunks now, but I could easily collapse it into a single commit if you want.

Its up to you, but due to the large amount of changes you might as well make a single commit and be done with it

**Comment by scottwittenburg** [comment]
You mentioned you tried using the slave approach and it worked for you:

> I Tried update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 100 --slave /usr/bin/g++ g++ /usr/bin/g++-11 in the ubuntu image and it seems to work.

Here's what I see:

```
$ docker run --rm -ti ecpe4s/ubuntu20.04:23.02
root@9bfae9858561:/# apt-get update
Hit:1 http://archive.ubuntu.com/ubuntu focal InRelease
Get:2 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:3 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Get:5 http://security.ubuntu.com/ubuntu focal-security/multiverse amd64 Packages [37.5 kB]
Get:6 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [2,529 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [3,346 kB]
Get:8 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages [2,866 kB]
Get:9 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages [1,070 kB]   
Get:10 http://archive.ubuntu.com/ubuntu focal-updates/restricted amd64 Packages [2,671 kB]    
Get:11 http://archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages [1,369 kB]
Get:12 http://archive.ubuntu.com/ubuntu focal-updates/multiverse amd64 Packages [40.3 kB]
Get:13 http://archive.ubuntu.com/ubuntu focal-backports/main amd64 Packages [55.2 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal-backports/universe amd64 Packages [28.6 kB]
Fetched 14.3 MB in 2s (6,591 kB/s)                           
Reading package lists... Done
root@9bfae9858561:/# update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 100 --slave /usr/bin/g++ g++ /usr/bin/g++-11
update-alternatives: error: alternative g++ can't be slave of gcc: it is a master alternative
```

Let me know if it's obvious to you what I'm doing wrong here, but I just copy-pasted what you said worked for you.

**Comment by scottwittenburg** [comment]
> > I'm not sure it's worth the effort to try to break it up into logical chunks now, but I could easily collapse it into a single commit if you want.
> 
> Its up to you, but due to the large amount of changes you might as well make a single commit and be done with it

Yeah ok, once you're happy with the changes (and everything is still working), I'll squash to a single commit and push one last time.

**Comment by scottwittenburg** [comment]
[5b4dc07](https://github.com/ornladios/ADIOS2/pull/3631/commits/5b4dc0748ce790bd344bd5a6088de13ec0402ef5) adds a `clang6` build, since it was easy to add, so we'll see how it fares in CI.  But if you prefer I can also save that commit for a follow-on PR.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake:38]: Is this needed?

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc8-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc8-mpi.cmake:38]: same

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc9-mpi.cmake
Comment (vicentebolea) [inline scripts/ci/cmake-v2/ci-ubuntu20.04-gcc9-mpi.cmake:38]: same here

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Looks great, before squashing everything into a single commit add a commit with the changes for `source/adios2/engine/bp5/BP5Reader.cpp`.

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake
Comment (scottwittenburg) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake:38]: I'm not sure whether it's needed.  It was present in about a third of the initial cache files in `scripts/ci/cmake-v2` before this, but I can't see any rhyme or reason to it.  Maybe some builds just behaved in a flaky manner without that limitation.  Should I try to just remove it and see?

**Comment by scottwittenburg** [comment]
> before squashing everything into a single commit add a commit with the changes for source/adios2/engine/bp5/BP5Reader.cpp

Should I separate out the change to `DataManWriter.cpp` in that same commit as well?

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake
Comment (vicentebolea) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake]: @scottwittenburg

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake
Comment (vicentebolea) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-gcc11-mpi.cmake:38]: I see, yes there has been certain level of tuning/tweaking with those params. Maybe remove and see is best here

**Comment by vicentebolea** [comment]
> > before squashing everything into a single commit add a commit with the changes for source/adios2/engine/bp5/BP5Reader.cpp
> 
> Should I separate out the change to `DataManWriter.cpp` in that same commit as well?

Yep

File: scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake
Comment (scottwittenburg) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake]: Do you want `Ninja` for all these builds?  In `master` I see:

```
$ grep -irn "unix makefiles" scripts/ci
scripts/ci/cmake-v2/ci-el8-oneapi-serial.cmake:25:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-ascent-nvhpc.cmake:28:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-gcc8-mpi.cmake:32:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-nvhpc222-mpi.cmake:33:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-oneapi-mpi.cmake:32:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-gcc8-serial.cmake:25:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-icc-mpi.cmake:32:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-common.cmake:15:  if(CTEST_CMAKE_GENERATOR STREQUAL "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-icc-serial.cmake:25:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
scripts/ci/cmake-v2/ci-el8-nvhpc222-serial.cmake:26:set(CTEST_CMAKE_GENERATOR "Unix Makefiles")
```

As with the ctest `PARALLEL_LEVEL`, I'm not sure I see the pattern as to when make is used vs ninja.  Possibly just to test both?

Comment (vicentebolea) [reply scripts/ci/cmake-v2/ci-ubuntu20.04-gcc10-mpich.cmake]: > Do you want Ninja for all these builds? In master I see:

The ones that were using Makefile you can leave it to keep using Makefile; new builds try to use Ninja.

Same thing for the parallel_level 1.

Note that mpich build do not use Makefile or parallel_level 1 in master.

**Review by vicentebolea — APPROVED** [review]
The two failing test for mpich are not related to this PR.

File: scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-base
Comment (vicentebolea) [inline scripts/ci/images-v2/Dockerfile.ci-spack-ubuntu20.04-base:21]:
```suggestion
    spack -e adios2-ci-ompi add ${ADIOS_SPEC_OPENMPI} && \
```

Will renaming this across the PR be a big change?

**Review by vicentebolea — CHANGES_REQUESTED** [review]
followup from previous review; everything else good; lets try ghcr instead of docker.

**Comment by scottwittenburg** [comment]
@vicentebolea Do you want to take one last look before merging?  I think I've addressed everything we discussed.

File: scripts/ci/images-v2/build-ubuntu.sh
Comment (vicentebolea) [inline scripts/ci/images-v2/build-ubuntu.sh:20]:
```suggestion
```

No need as we will use GHCR

**Review by vicentebolea — CHANGES_REQUESTED** [review]
Lookgs good small last detail

**Review by vicentebolea — APPROVED** [review]
Ready to merge; If you cant merge let me know.

Good job there :)

**Comment by scottwittenburg** [comment]
Thanks @vicentebolea!  It seems that I cannot merge, but I don't think it's about permissions, as it seems I could push the "Enable auto-merge" button to "Automatically merge when all requirements are met".  Is it possible the failing `mpich` test is preventing merge?

**Comment by vicentebolea** [comment]
@scottwittenburg yeah, I have also rebase it, lets merge it after passing CI

**Pull request merged**
